### PR TITLE
Add Account Registration

### LINKED ISSUE #6: User Story - Account Registration
# Story Description

As a user, I would like to be able to create an account on the website so I can possess resources that are unique to me (e.g. a profile).

# Acceptance Criteria

- Account information should be stored inside a database.
  - Passwords **must** be hashed.
- Account information should be validated (duplicate emails, blank names, etc).
- API route `POST /api/v1/users/register`.
- UI.

# Additional Info

N/A

### PR DESCRIPTION
**Pull request by miosenpai**
This PR completes #6 with the following changes:

- Backend connection to PostgreSQL has been set up.
- User (authentication use only) and Profile entities were added.
- `POST /api/v1/users/register` has been added.
  - Corresponding tests included as well.
- Automatic frontend type generation has been set up.
  - Built on top of existing backend OpenAPI documentation.
- Frontend registration page was added.
  - We will consider login after successful registration during #7.
- The README file was revamped slightly for tidiness.

Note this PR really has 887 lines added, the rest out of the 2439 total are from auto generated DB migration files (as indicated by the comments in those files) and Yarn's lock file.

Due to a limitation within the Identity module of ASP.NET Core, unused attributes (e.g. 2FA) show up in the auto generated DB migration files (details [here](https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/10#issuecomment-1025451032)).

### CODE REVIEW

File: Buddies.Web/src/pages/register.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                || errorField === 'email'
+                || errorField === 'password') {
+              const errorMsg = problems.errors![errorField].join('\n');
+              formMethods.setError(errorField, { message: errorMsg });
+            }
+          });
+        }
+      }
+    }
+  };
```
Comment (andy9775) [inline Buddies.Web/src/pages/register.tsx]: I recommend having an api class which handles all of the HTTP logic. That way it's centralized and the core logic is easier to test. Furthermore, the logic for posting and getting data is 90% the same between requests - only thing that changes is the parameters, path and response. Using generics you can simplify a lot of the code.

File: Buddies.Web/src/pages/index.tsx [hunk 2]
```diff
@@ -2,6 +2,9 @@ import type { NextPage } from 'next';
 import Head from 'next/head';
 import Typography from '@mui/material/Typography';
 
+/**
+ * Home page.
+ */
```
Comment (andy9775) [inline Buddies.Web/src/pages/index.tsx]:
```suggestion
```

File: Buddies.Web/src/pages/_app.tsx [hunk 3]
```diff
@@ -7,7 +7,11 @@ import useMediaQuery from '@mui/material/useMediaQuery';
 import { useMemo } from 'react';
 import { createTheme, ThemeProvider } from '@mui/material/styles';
 import CssBaseline from '@mui/material/CssBaseline';
+import Layout from '../components/Layout';
 
+/**
+ * Custom Next.js application.
```
Comment (andy9775) [inline Buddies.Web/src/pages/_app.tsx]: Should be more descriptive, or none at all.

File: Buddies.Web/src/components/Layout.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import Box from '@mui/material/Box';
+import React from 'react';
+import Navbar from './Navbar';
+
+/**
+ * Site layout component.
```
Comment (andy9775) [inline Buddies.Web/src/components/Layout.tsx]: Should be more descriptive or none at all

File: Buddies.Tests/User/RegisterTests.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+        var responseBody = response.Content.ReadFromJsonAsync<JsonDocument>().Result;
+        Assert.True(responseBody?.RootElement.GetProperty("errors").TryGetProperty("firstName", out _));
+        Assert.True(responseBody?.RootElement.GetProperty("errors").TryGetProperty("lastName", out _));
+    }
+
+    [Fact]
+    public void TestUserRegistered()
+    {
+        var request = new RegisterRequest("Akari", "Akaza", "akari@akaza.moe", "Pwd^123");
```
Comment (andy9775) [inline Buddies.Tests/User/RegisterTests.cs:71]: Should we allow only UofT emails?

File: Buddies.Tests/User/RegisterTests.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        
+        var response = _client.PostAsJsonAsync("/api/v1/users/register", request).Result;
+        Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+        var responseBody = response.Content.ReadFromJsonAsync<JsonDocument>().Result;
+        Assert.True(responseBody?.RootElement.GetProperty("errors").TryGetProperty("firstName", out _));
+        Assert.True(responseBody?.RootElement.GetProperty("errors").TryGetProperty("lastName", out _));
+    }
+
+    [Fact]
+    public void TestUserRegistered()
```
Comment (andy9775) [inline Buddies.Tests/User/RegisterTests.cs:69]: Add test for duplicate email address (user already exists).

File: Buddies.API/Migrations/ApiContextModelSnapshot.cs [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+                    b.Property<string>("PhoneNumber")
+                        .HasColumnType("text");
+
+                    b.Property<bool>("PhoneNumberConfirmed")
+                        .HasColumnType("boolean");
+
+                    b.Property<string>("SecurityStamp")
+                        .HasColumnType("text");
+
+                    b.Property<bool>("TwoFactorEnabled")
```
Comment (andy9775) [inline Buddies.API/Migrations/ApiContextModelSnapshot.cs:121]: Are we implementing two factor at this time?

**Review by andy9775 — CHANGES_REQUESTED** [review]
Just some feedback after the first pass:

- This should have been broken down into a few pr's. Connect the DB, setup the models, setup the registration page, etc. It makes it easier to review if things are compact and related.
- Only add the related db columns. e.g. if two factor isn't being implemented in this PR, don't add the column. Each PR should include all the columns and code they need to implement the feature in question.
- Avoid obvious documentation/comments

File: Buddies.Tests/User/RegisterTests.cs (hunk 5 above)
Comment (miosenpai) [reply Buddies.Tests/User/RegisterTests.cs:71]: According to the original project proposal, the site was not meant to be internal to UofT.

**Comment by miosenpai** [comment]
> Only add the related db columns. e.g. if two factor isn't being implemented in this PR, don't add the column. Each PR should include all the columns and code they need to implement the feature in question.

So I can [prevent those fields from entering the DB](https://stackoverflow.com/questions/25559878/asp-net-identity-remove-column-from-aspnetusers-table) but the fields will still be present within the C# code since they are on the `IdentityUser` base class from the ASP.NET Core Identity library which we are highly recommended to extend from when using said library, this could cause side effects on library classes such as `UserManager` or `SignInManager` which may rely on those fields behind the scenes thus I don't feel particularly safe about deleting schema fields here. Now the Identity library does support custom "user" classes that do not inherit from `IdentityUser` but this approach would involve reimplementing the aforementioned library classes which somewhat negates the benefits of using it in the first place.

File: Buddies.Tests/User/RegisterTests.cs (hunk 6 above)
Comment (miosenpai) [reply Buddies.Tests/User/RegisterTests.cs:69]: Added.

**Review by TianYueDong — APPROVED** [review]
Looks good!

File: Buddies.API/Migrations/ApiContextModelSnapshot.cs (hunk 7 above)
Comment (miosenpai) [reply Buddies.API/Migrations/ApiContextModelSnapshot.cs:121]: See [comment](https://github.com/GDSCUTM-CommunityProjects/Buddies/pull/10#issuecomment-1025451032).

File: Buddies.Web/src/pages/register.tsx (hunk 1 above)
Comment (miosenpai) [reply Buddies.Web/src/pages/register.tsx]: As discussed during weekly meeting, actual API call has been refactored into separate file for maintainability but error handling code is kept in component as it relies on local state (e.g. `setError`).

**Review by andy9775 — APPROVED** [review]
LGTM!
